// @factory/Factory.js

import { Render } from '@factory/Render'

const defaultDNA = {
    "headColor" : 10,
    "mouthColor" : 13,
    "eyesColor" : 96,
    "earsColor" : 10,
    //Cattributes
    "eyesShape" : 1,
    "decorationPattern" : 1,
    "decorationMidColor" : 13,
    "decorationSidesColor" : 13,
    "animation" : 0,
    "lastNum" : 1
    }
    
export const defaultDragon = {
    id: "",
    dna: defaultDNA,
    gen: "0"
}

export class Factory extends Render {

    getDefaultKittie() {
        try {
            this.render(defaultCat)
            this.updateSliders(defaultCat.dna)
        }
        catch (error) {
            console.log(`Error In getDefaultKittie(): ${error}`)
        }
    }


    getRandomKittie() {
        try {
            const randomCat = {
                id: "",
                gen: "",
                dna: {
                    "headColor": getRandomIntegerBetween(10, 98),
                    "mouthColor": getRandomIntegerBetween(10, 98),
                    "eyesColor": getRandomIntegerBetween(10, 98),
                    "earsColor": getRandomIntegerBetween(10, 98),
                    //Cattributes
                    "eyesShape": getRandomIntegerBetween(0, 9),
                    "decorationPattern": getRandomIntegerBetween(0, 9),
                    "decorationMidColor": getRandomIntegerBetween(10, 98),
                    "decorationSidesColor": getRandomIntegerBetween(10, 98),
                    "animation": getRandomIntegerBetween(0, 9),
                    "lastNum": getRandomIntegerBetween(0, 9)
                }
            }

            this.render(randomCat)
            this.updateSliders(randomCat.dna)
        }
        catch (error) {
            console.log(`Error In getRandomKittie(): ${error}`)
        }
    }

    
    getHtmlForKitty(dragon) {
    try {
        let html = `
      <div class="col-lg-4" id="kitty${dragon.id}" align="left">
        <div class="row col-sm-12">

            <div class="col-sm-6" id="catDNA" align="left">
                <span class="badge bg-dark ml-2">
                    
                    DNA:
                    <!-- Colors -->
                    <span id="dnabody"></span>
                    <span id="dnamouth"></span>
                    <span id="dnaeyes"></span>
                    <span id="dnaears"></span>

                    <!-- Cattributes -->
                    <span id="dnashape"></span>
                    <span id="dnadecoration"></span>
                    <span id="dnadecorationMid"></span>
                    <span id="dnadecorationSides"></span>
                    <span id="dnaanimation"></span>
                    <span id="dnaspecial"></span>
                </span>
            </div> 

            <div class="col-sm-6" align="right">
                <span class="badge bg-light text-dark">
        
                        GEN: <span id="catGenNum">0</span>
                
                </span>
            </div>
        </div>

        <div class="catBox light-b-shadow">
        
            <div class="cat">
                <div class="cat__ear">
                    <div id="leftEar" class="cat__ear--left">
                        <div class="cat__ear--left-inside"></div>
                    </div>
                    <div id="rightEar" class="cat__ear--right">
                        <div class="cat__ear--right-inside"></div>
                    </div>
                </div>

                <div id="head" class="cat__head">
                    <div id="midDot" class="cat__head-dots">
                        <div id="leftDot" class="cat__head-dots_first"></div>
                        <div id="rightDot" class="cat__head-dots_second"></div>
                    </div>
                    <div id="eyes" class="cat__eye">
                        <div class="cat__eye--left">
                            <span class="pupil-left"></span>
                        </div>
                        <div class="cat__eye--right">
                            <span class="pupil-right"></span>
                        </div>
                    </div>
                    <div id="nose" class="cat__nose"></div>

                    <div class="cat__mouth-contour"></div>
                    <div class="cat__mouth-left"></div>
                    <div class="cat__mouth-right"></div>

                    <div id="leftWhiskers" class="cat__whiskers-left"></div>
                    <div id="rightWhiskers" class="cat__whiskers-right"></div>
                </div>

                <div class="cat__body">

                    <div class="cat__chest"></div>
                    <div class="cat__chest_inner"></div>

                    <div id="leftRearPaw" class="cat__paw-left_rear"></div>
                    <div id="leftFrontPaw" class="cat__paw-left_front"></div>

                    <div id="rightFrontPaw" class="cat__paw-right_front"></div>
                    <div id="rightRearPaw" class="cat__paw-right_rear"></div>

                    <div id="tail" class="cat__tail"></div>
                </div>
            </div>
            <br>
        
        </div>
      </div>`;

        document.getElementById("dragonGrid").innerHTML += html;
        
    } catch (error) {
        console.log("In getHtmlForKitty(id): " + error);
    }
}

}


function getRandomIntegerBetween(low, high){
    try {
      if (low<0) throw "Error: Only supports positive integers - low is negative!"
      if (low>high) throw "Error: low is greater than high!"
  
      const rangeInclusive = high - low + 1 
      const RandomValue = Math.floor(Math.random() * rangeInclusive) + low
  
      return RandomValue
    }
    catch(error) {
      throw(`In getRandomIntegerBetween(${low}, ${high}): ${error}`)
    }
  }
