// @factory/Settings.js

import { allColors } from '@factory/Colors'

export class Settings {
    constructor() {  
        this.colors = allColors()
        this.eyeVariations = [
            {
              "name" : "Big Round Eyes",
              "setEyesFunc" : this.normalEyes
            },
            {
              "name" : "Look Down",
              "setEyesFunc" : this.eyesType1
            },
            {
                "name" : "Look Up",
                "setEyesFunc" : this.eyesType2
            },
            {
                "name" : "Narrow pupils",
                "setEyesFunc" : this.eyesType3
            },
            {
                "name" : "Narrow pupils, looking left",
                "setEyesFunc" : this.eyesType4
            },
            {
                "name" : "Narrow pupils, looking right",
                "setEyesFunc" : this.eyesType5
            },
            {
                "name" : "Cross-eyed",
                "setEyesFunc" : this.eyesType6
            },
            {
                "name" : "Lazy left-eye",
                "setEyesFunc" : this.eyesType7
            },
            {
                "name" : "Lazy right-eye",
                "setEyesFunc" : this.eyesType8
            },
            {
                "name" : "Two lazy eyes",
                "setEyesFunc" : this.eyesType9
            }
          ]
        
         this.decorationVariations = [
            {
                "name" : "Short stripes downs",
                "setDecorationFunc" : this.normaldecoration
            },
            {
                "name" : "Long stripes up",
                "setDecorationFunc" : this.patternType1
            },
            {
                "name" : "Angle stripes",
                "setDecorationFunc" : this.patternType2
            },
            {
                "name" : "'Bald' patch",
                "setDecorationFunc" : this.patternType3
            },
            {
                "name" : "3-prong Leaf",
                "setDecorationFunc" : this.patternType4
            },      
            {
                "name" : "Stripe with side blobs",
                "setDecorationFunc" : this.patternType5
            },
            {
                "name" : "Two stripes",
                "setDecorationFunc" : this.patternType6
            },      
            {
                "name" : "One strip",
                "setDecorationFunc" : this.patternType7
            },      
            {
                "name" : "Spikey hair",
                "setDecorationFunc" : this.patternType8
            },
            {
                "name" : "No pattern",
                "setDecorationFunc" : this.patternType9
            }
        ]
        
         this.animationVariations = [
            {
                "name" : "Roll head",
                "setAnimationFunc" : this.animationType1
            },
            {
                "name" : "Swish tail",
                "setAnimationFunc" : this.animationType2
            },
            {
                "name" : "Wiggle left ear",
                "setAnimationFunc" : this.animationType3
            },
            {
                "name" : "Wiggle right ear",
                "setAnimationFunc" : this.animationType4
            },
            {
                "name" : "Wiggle both ears",
                "setAnimationFunc" : this.animationType5
            },      
            {
                "name" : "Alert ears",
                "setAnimationFunc" : this.animationType6
            },
            {
                "name" : "Twitch nose & whiskers",
                "setAnimationFunc" : this.animationType7
            },      
            {
                "name" : "Wandering eyes",
                "setAnimationFunc" : this.animationType8
            },
            {
                "name" : "Standing up",
                "setAnimationFunc" : this.animationType9
            },
            {
                "name" : "Hyperactive",
                "setAnimationFunc" : this.animationType10
            }
        ]
    
        }  

    headColor(code, idCat="") {
        $(`${idCat} .cat__head, ${idCat} .cat__chest`).attr('style','background:#' + this.colors[code])
        $(`${idCat} #dnabody`).html(code)   // Update DNA display (below the cat)
    }
    
     mouthChestTailColor(code, idCat="") {
        $(`${idCat} .cat__mouth-contour, ${idCat} .cat__chest_inner, ${idCat} .cat__tail`).css(
            'background', '#' + this.colors[code]
        )
        $(`${idCat} #dnamouth`).html(code)
    }
    
     eyesColor(code, idCat="") {
        $(`${idCat} [class^="pupil-"]`).css('background', '#' + this.colors[code])  
        $(`${idCat} #dnaeyes`).html(code)
    }
    
     earsPawsColor(code, idCat="") {
        $(`${idCat} [id$="Ear"], ${idCat} [class^="cat__paw"]`).css('background', '#' + this.colors[code])
        $(`${idCat} #dnaears`).html(code)
    }
    
     innerDecorationColor(code, idCat="") {
        $(`${idCat} .cat__head-dots`).css('background', '#' + this.colors[code]) 
        $(`${idCat} #dnadecorationMid`).html(code)
    }
    
     outerDecorationColor(code, idCat="") {
        $(`${idCat} [class^="cat__head-dots_"]`).css('background', '#' + this.colors[code])     
        $(`${idCat} #dnadecorationSides`).html(code)
    }
    
     eyeVariation(num, idCat="") {
        this.normalEyes(idCat) //Reset eyes
        $(`${idCat} #dnashape`).html(num)   // Update DNA display (below cat)
        this.eyeVariations[num].setEyesFunc(idCat)
    }
    
     normalEyes(idCat) {
        // Reset eye lids to fully open
        $(`${idCat} .cat__eye`).find('span').css('border', 'none')
    
        // Reset pupil to round and centered
        $(`${idCat} .cat__eye`).find('span').css('width', '42px') 
        $(`${idCat} .pupil-left`).css('left', '42px')    
        $(`${idCat} .pupil-right`).css('left', '166px') 
    }
    
     eyesType1(idCat) {  // Look down
        $(`${idCat} .cat__eye`).find('span').css('border-top', '15px solid')
    }
    
     eyesType2(idCat) {  // Look up
        $(`${idCat} .cat__eye`).find('span').css('border-bottom', '15px solid')
    }
    
     eyesType3(idCat) {  // Narrow pupils (straight ahead)
        $(`${idCat} .cat__eye`).find('span').css('width', '22px')  // Wide-42, Narrow-22 
        $(`${idCat} .pupil-left`).css('left', '52px')    // Recentre left & right pupils (ie. compensate for
        $(`${idCat} .pupil-right`).css('left', '176px')  // 20px width change by moving pupils left +10px)
    }
    
     eyesType4(idCat) {  // Narrow pupils looking left
        this.eyesType3(idCat) //Narrow pupils 
        $(`${idCat} .pupil-left`).css('left', '42px')    
        $(`${idCat} .pupil-right`).css('left', '166px') 
    }
    
     eyesType5(idCat) {    // Narrow pupils looking right
        this.eyesType3(idCat) //Narrow pupils 
        $(`${idCat} .pupil-left`).css('left', '62px')
        $(`${idCat} .pupil-right`).css('left', '186px')
    }
    
     eyesType6(idCat) {  // Cross-eyed
        this.eyesType3(idCat) //Narrow pupils
        $(`${idCat} .pupil-left`).css('left', '62px') 
        $(`${idCat} .pupil-right`).css('left', '166px')
    }
    
     eyesType7(idCat) {  // Lazy left-eye
        this.eyesType3(idCat) //Narrow pupils
        $(`${idCat} .pupil-left`).css('left', '42px')
    }
    
     eyesType8(idCat) {  // Lazy right eye
        this.eyesType3(idCat) //Narrow pupils
        $(`${idCat} .pupil-right`).css('left', '186px')
    }
    
     eyesType9(idCat) {  // Two lazy eyes
        this.eyesType3(idCat) //Narrow pupils 
        $(`${idCat} .pupil-left`).css('left', '42px') 
        $(`${idCat} .pupil-right`).css('left', '186px')
    }
    
    
     decorationVariation(num, idCat="") {
        this.normaldecoration(idCat) //Reset decoration
        $(`${idCat} #dnadecoration`).html(num)   // Update DNA display (below cat)
        this.decorationVariations[num].setDecorationFunc(idCat)
    }
    
     normaldecoration(idCat) {
        //Remove all style from other decorations (to reset)
        $(`${idCat} .cat__head-dots`).css({ "transform": "rotate(0deg)", "height": "48px", "width": "14px",
                                            "top": "1px", "border-radius": "0 0 50% 50%", "left": "101" })
        $(`${idCat} .cat__head-dots_first`).css({ "transform": "rotate(0deg)", "height": "35px", "width": "14px",
                                            "top": "1px", "border-radius": "50% 0 50% 50%", "left": "-20px"})
        $(`${idCat} .cat__head-dots_second`).css({ "transform": "rotate(0deg)", "height": "35px", "width": "14px",
                                            "top": "1px", "border-radius": "0 50% 50% 50%", "left": "20px" })
    }
    
     patternType1(idCat) {  // Long stripes up
        $(`${idCat} .cat__head-dots`).css({ "transform": "rotate(180deg)", "height": "95px"})
        $(`${idCat} [class^="cat__head-dots_"]`).css('height', '80px')
    }
    
     patternType2(idCat) {  //Angle stripes
        $(`${idCat} .cat__head-dots`).css({ "transform": "rotate(-25deg)", "top":"-6px"})
        $(`${idCat} .cat__head-dots_first`).css('top','-5px')
        $(`${idCat} .cat__head-dots_second`).css({ "top":"8px", "border-radius":"0px 40% 50% 50%"})
    }
    
     patternType3(idCat) {  //'Bald' patch
        $(`${idCat} .cat__head-dots`).css({ "height":"40px", "width":"100px", "left": "60px",
                                            "border-radius": "50% 50% 30px 30px" })
        $(`${idCat} [class^="cat__head-dots_"]`).css('width', '0px')
    }
    
     patternType4(idCat) {  //3-prong leaf
        $(`${idCat} .cat__head-dots_first`).css({ "transform":"rotate(36deg)", "border-radius":"50% 0px 50% 50%" })
        $(`${idCat} .cat__head-dots_second`).css({ "transform":"rotate(-36deg)", "border-radius":"0px 50% 50% 50%" })
    }
    
     patternType5(idCat) {  //Stripe with side blobs
        $(`${idCat} .cat__head-dots`).css({ "width":"20px", "left": "98px" })
        $(`${idCat} .cat__head-dots_first`).css({ "width":"35px", "left":"-38px", "border-radius":"25px 0px 50% 50%"})
        $(`${idCat} .cat__head-dots_second`).css({ "width":"35px", "left":"23px", "border-radius":"0px 25px 50% 50%" })
    }
    
     patternType6(idCat) {  //Two stripes
        $(`${idCat} .cat__head-dots`).css('width', '0px')
    }
    
     patternType7(idCat) {  //One stripe 
        $(`${idCat} [class^="cat__head-dots_"]`).css('width', '0px')
    }
    
     patternType8(idCat) {  //Spikey hair
        $(`${idCat} .cat__head-dots`).css({ "transform": "rotate(180deg)", "top": "-40px" })
    }
    
     patternType9(idCat) {  //No pattern
        $(`${idCat} .cat__head-dots`).css('width', '0px')
        $(`${idCat} [class^="cat__head-dots_"]`).css('width', '0px')
    }
    
    
     animationVariation(num, idCat="") {
        this.removeAllAnimations(idCat) //Reset animations
        $(`${idCat} #dnaanimation`).html(num)   // Update DNA display (below cat)
        this.animationVariations[num].setAnimationFunc(idCat)
    }
    
    
     removeAllAnimations(idCat) { 
        $(`${idCat} #head`).attr("class", "cat__head")
        $(`${idCat} #leftEar`).attr("class", "cat__ear--left")
        $(`${idCat} #rightEar`).attr("class", "cat__ear--right")
        $(`${idCat} #tail`).attr("class", "cat__tail")
    
        $(`${idCat} #nose`).attr("class", "cat__nose")
        $(`${idCat} #leftWhiskers`).attr("class", "cat__whiskers-left")
        $(`${idCat} #rightWhiskers`).attr("class", "cat__whiskers-right")
    
        $(`${idCat} #eyes`).attr("class", "cat__eye")
        $(`${idCat} #cat`).attr("class", "cat")
    
        $(`${idCat} #leftFrontPaw`).attr("class", "cat__paw-left_front")
        $(`${idCat} #rightFrontPaw`).attr("class", "cat__paw-right_front")
        $(`${idCat} #leftRearPaw`).attr("class", "cat__paw-left_rear")
        $(`${idCat} #rightRearPaw`).attr("class", "cat__paw-right_rear")
    }
    
     animationType1(idCat) {  // Roll Head
        $(`${idCat} #head`).addClass("movingHead")
        $(`${idCat} #leftEar`).addClass("movingLeftEar")
        $(`${idCat} #rightEar`).addClass("movingRightEar")
    }
    
     animationType2(idCat) {  // Swish Tail
        $(`${idCat} #tail`).addClass("movingTail")
    }
    
     animationType3(idCat) {  // Wiggle left ear
        $(`${idCat} #leftEar`).addClass("wiggleLeftEar")
    }
    
     animationType4(idCat) {  // Wiggle right ear
        $(`${idCat} #rightEar`).addClass("wiggleRightEar")
    }
    
     animationType5(idCat) {  // Wiggle both ears
        $(`${idCat} #leftEar`).addClass("wiggleLeftEar")
        $(`${idCat} #rightEar`).addClass("wiggleRightEar")
    }
    
     animationType6(idCat) {  // Alert ears
        $(`${idCat} #leftEar`).addClass("alertingLeftEar")
        $(`${idCat} #rightEar`).addClass("alertingRightEar")
    }
    
     animationType7(idCat) {  // Twitch nose & whiskers
        $(`${idCat} #nose`).addClass("twitchingNose")
        $(`${idCat} #leftWhiskers`).addClass("twitchingLeftWhiskers")
        $(`${idCat} #rightWhiskers`).addClass("twitchingRightWhiskers")
    }
    
     animationType8(idCat) {  // Wandering eyes
        $(`${idCat} #eyes`).addClass("wanderingEyes")
    }
    
     animationType9(idCat) {  // Stand up
        $(`${idCat} #cat`).addClass("catRising")
        $(`${idCat} #leftRearPaw, ${idCat} #rightRearPaw`).addClass("extendingRearLeg")
        $(`${idCat} #leftFrontPaw, ${idCat} #rightFrontPaw`).addClass("raisingFrontLeg")
    }
    
     animationType10(idCat) {  // Hyperactive
        this.animationType1(idCat)     // Roll head
        this.animationType2(idCat)     // Swish Tail
        this.animationType5(idCat)    // Wiggle both ears
        this.animationType7(idCat)    // Twitch nose & whiskers
        this.animationType8(idCat)   // Wandering eyes
        this.animationType9(idCat)   // Standing up
    }
}
