// @storage/contract.js
import { writable } from 'svelte/store';
import { contractsAbi } from "@contracts/contractsAbi";

export const kittyAddress = "0x996693a83BCe641A3ac6A092A0f66020BCDb5574";
export const marketplaceAddress = "0x98a11542C588b14912A522B6afC17e9678cC8f9f";

export const balances = writable({ contract: 0,user: 0 });

export const contractInfo = async () => {
    
    const web3 = await Moralis.Web3.enable();
    let account = web3.currentProvider.selectedAddress;
    let kittyContract = new web3.eth.Contract(contractsAbi.kittyContract, kittyAddress, {
        from: account,
    });
    let marketplace = new web3.eth.Contract(contractsAbi.marketplace, marketplaceAddress, {
        from: account,
    });

    let interfaces = {
        kittyContract,
        marketplace
    }
    
    let contractData = {              
        account: account,
        interfaces : interfaces
    } 
    return contractData

}

export const updateBalances = async (dappBalance,userBalance) => {

    balances.update(value =>{
        value = { contract: dappBalance,user: userBalance }      
        return value
    }  )  

}
