// @storage/contract.js
import { writable } from 'svelte/store';
import { contractsAbi } from "@contracts/contractsAbi";
//old
//0xA959385EA23dFF9DaA79cb5F95Baf3398EBc8EAE
export const dragonContract = "0x7744002e44CfA783F2c9787B5E0B6a7f03003c87";
//old
//0x2Ea84C95C5B16Fa42254EB195Ef72d8E8D015a9a
export const marketplaceAddress = "0xf2E73E7Ebc8fbdF4155569D122978478B1a36D20";
export const balances = writable({ contract: 0,user: 0 });
export const contractInfo = async () => {
    
    const web3 = await Moralis.Web3.enable();
    let account = web3.currentProvider.selectedAddress;

    let dragon_instance = new web3.eth.Contract(contractsAbi.dragonToken, dragonContract, {
        from: account,
    });

    let marketplace_instance = new web3.eth.Contract(contractsAbi.marketplace, marketplaceAddress, {
        from: account,
    });

    let interfaces = {
        dragon:dragon_instance,
        marketplace: marketplace_instance
    }

    let address = {
        dragon:dragonContract,
        marketplace: marketplaceAddress
    }
    
    let contractData = {              
        address:address,
        account: account,
        interfaces : interfaces
    } 
    return contractData

}

export const updateBalances = async (dappBalance,userBalance) => {

    balances.update(value =>{
        value = { contract: dappBalance,user: userBalance }      
        return value
    }  )  

}
