// @storage/dragon.js
import { Factory } from "@factory/Factory";
import { writable } from 'svelte/store';

const FactoryClass = new Factory();

export function offer_format(dragon){

    const dragonObj = {
        id: dragon.tokenId,
        dna: FactoryClass.dnaFromGenes(dragon.genes),
        gen: dragon.generation,
        displayDna:true,
        displayInfo:false,
        displayAttributes:false,
    };

    return dragonObj

}

export function user_format(dragon){

    const dragonObj = {
        id: dragon.tokenId,
        dna: FactoryClass.dnaFromGenes(dragon.genes),
        gen: dragon.generation,
        displayDna:true,
        displayInfo:true,
        displayAttributes:true,
    };

    return dragonObj

}

export function marketplace_format(dragon) {

    const dragonObj = {
        id: dragon.tokenId,
        dna: FactoryClass.dnaFromGenes(dragon.genes),
        gen: dragon.generation,
        displayDna: true,
        displayInfo: true,
        displayAttributes: true,
    };

    return dragonObj;
}
