// @storage/dragon.js
import { Factory } from "@factory/Factory";
import { writable } from 'svelte/store';

export const showDragon = writable({ show:false });

export function displayAll(){
    showDragon.update(value => {
        value = { show:false }
        return value;
    });   
}

export function displaySingle(){
    showDragon.update(value => {
        value = { show:true }
        return value;
    });   
}

const FactoryClass = new Factory();



export function marketplace_dragon(dragon){

    const dragonObj = {
        id: dragon.tokenId,
        dna: FactoryClass.dnaFromGenes(dragon.genes),
        gen: dragon.generation,
        mumId:dragon.mumId,
        dadId:dragon.dadId,
        displayOffer:true,
        displayDna:true,
        displayInfo:false,
        displayAttributes:false,
    };

    return dragonObj

}

export function cave_dragon(dragon){

    const dragonObj = {
        id: dragon.tokenId,
        dna: FactoryClass.dnaFromGenes(dragon.genes),
        gen: dragon.generation,
        mumId:dragon.mumId,
        dadId:dragon.dadId,
        displayDna:true,
        displayInfo:false,
        displayAttributes:false,
    };

    return dragonObj

}
