// @controllers/user/login_validation.js
import { sequelize } from '@models/database/squalized'
import { User } from '@models/user/user_model'
import { createSession } from '@storage/userData'


export let login_validation = async (req, res) => {

    const { user, password } = req.body
    const errHandler = (err) => {
        console.log("Error: ", err)
    }
    var results = {}
    try {

        await sequelize.authenticate();
        const userData = User.build({ username: user, password: password });
        const userIn = await User.findAll({
            where: {
                username: userData.username,
                password: userData.password
            },
            limit: 1,
            raw: true
        }).catch(errHandler);
   
        if (userIn.length) {
            let dbUser = userIn[0]
            let userInfo = createSession(req, dbUser)
            results = { status: 200, msg: "Logged in success!", sessionData: userInfo }
         
        } else {
            results = { status: 203, msg: "Invalid user or credentials." }
        }
        return res.end(JSON.stringify(results));

    } catch (err) {
        errHandler(err)
    }

}
