// @helpers/helpers.js

import { userSession } from "@storage/userData";
import { getErrors } from "@helpers/errHandlers";


export async function postCall(url, params) {
try {
    const res = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(params),
    });
    const result = await res.json();
    return result

} catch(err){
    console.log(err)
}

}

export async function getCall(url) {
    try {
    const res = await fetch(url, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
    });
    const result = await res.json();
    return result

} catch(err){
    console.log(err)
}
}


//RETURN ALERT FOR A USER FORM
export function requestHanlder(result, successMsg,updateSession=false) {

    let alertData = {};

    if (result.status != 200) {
        alertData.content = getErrors(result.errors);
        alertData.alertType = "warning";
    } else {
        alertData.content = successMsg;
        alertData.alertType = "success";
        if(updateSession){
            userSession.update((value) => {
                return (value = result.sessionData);
            });
        }
       
    }
    return alertData;
}
